lib/peep/codegen.ex

```diff
@@ -39,33 +42,35 @@ defmodule Peep.Codegen do
     quote do
       def handle_event(event, measurements, metadata, _) do
         global_tags = global_tags()
-        global_tags_keys = global_tags_keys()
 
         %Persistent{
           events_to_metrics: %{^event => metrics},
           storage: {storage_mod, storage}
-        } = Persistent.fetch(unquote(peep_name))
-
-        for {metric, id} <- metrics do
-          %{
-            measurement: measurement,
-            tag_values: tag_values,
-            tags: tags,
-            keep: keep
-          } = metric
-
-          case keep?(keep, metadata) && fetch_measurement(measurement, measurements, metadata) do
-            value when is_number(value) ->
-              tag_values = Map.merge(global_tags, tag_values.(metadata))
-              tags = Map.new(global_tags_keys ++ tags, &{&1, Map.get(tag_values, &1, "")})
-              storage_mod.insert_metric(storage, id, metric, value, tags)
-
-            _ ->
-              nil
-          end
-        end
-
-        :ok
+        } = Persistent.fast_fetch(unquote(peep_name))
+
+        :lists.foreach(
+          fn {metric, id} ->
+            %{
+              measurement: measurement,
+              tag_values: tag_values,
+              tags: tags,
+              keep: keep
+            } = metric
+
+            if keep?(keep, metadata) do
+              case fetch_measurement(measurement, measurements, metadata) do
+                value when is_number(value) ->
+                  tag_values = tag_values.(metadata)
+                  tags = Map.merge(global_tags, Map.take(tag_values, tags))
+                  storage_mod.insert_metric(storage, id, metric, value, tags)
+
+                _ ->
+                  nil
+              end
+            end
+          end,
+          metrics
+        )
```

Comment:
I wonder whether it wouldn't be better to create separate handler for each metric in Telemetry instead of having single large handler that will then dispatch to respective metrics storage units.